Plugins~/Src/MeshSyncClientBlender/msblenMaterialsExportHelper.cpp

```diff
@@ -0,0 +1,497 @@
+	if (!node || node->type != NODE_REROUTE) {
+		return node;
+	}
+
+	auto tree = mat->nodetree;
+
+	for (auto link : list_range((bNodeLink*)tree->links.first)) {
+		if (link->tonode == node) {
+			return removeReroutes(link->fromnode, mat);
+		}
+	}
+
+	return nullptr;
+}
+
+bNode* getNodeConnectedToSocket(bNodeSocket* socket) {
+	if (socket->link) {
+		return socket->link->fromnode;
+	}
+
+	return nullptr;
+}
+
+// For mix shaders, we pass through the first input that has a connection.
+bNode* handleMixShader(const Material* mat, bNode* bsdf) {
+	if (!bsdf || bsdf->type != SH_NODE_MIX_SHADER)
+		return bsdf;
+
+	for (auto inputSocket : list_range((bNodeSocket*)bsdf->inputs.first)) {
+		if (STREQ(inputSocket->name, shaderIdentifier)) {
+			bNode* connectedBSDF = removeReroutes(getNodeConnectedToSocket(inputSocket), mat);
+			if (connectedBSDF)
+			{
+				return connectedBSDF;
+			}
+		}
+	}
+
+	return bsdf;
+}
+
+bool getBSDFAndOutput(const Material* mat, bNode*& bsdf, bNode*& output) {
+	// Find BSDF connected to an output node:
+	auto tree = mat->nodetree;
+	for (auto link : list_range((bNodeLink*)tree->links.first)) {
+		if (link->tonode &&
+			link->tonode->type == SH_NODE_OUTPUT_MATERIAL &&
+			STREQ(link->tosock->identifier, surfaceIdentifier)) {
+			bsdf = removeReroutes(link->fromnode, mat);
+
+			bsdf = handleMixShader(mat, bsdf);
+
+			output = link->tonode;
+			return bsdf && output;
+		}
+	}
+
+	return false;
+}
+
+bNodeSocket* getInputSocket(bNode* node, const char* socketName) {
+	for (auto inputSocket : list_range((bNodeSocket*)node->inputs.first)) {
+		if (STREQ(inputSocket->name, socketName)) {
+			return inputSocket;
+		}
+	}
+
+	return nullptr;
+}
+
+int msblenMaterialsExportHelper::exportTexture(const std::string& path, ms::TextureType type) const
+{
+	return m_texture_manager->addFile(path, type);
+}
+
+void msblenMaterialsExportHelper::exportPackedImages(ms::TextureType& textureType,
+	std::function<void(int textureId)> setTextureHandler,
+	Image* img) const
+{
+	for (auto imagePackedFile : list_range((ImagePackedFile*)img->packedfiles.first)) {
+		std::string name = img->id.name + 2; // +2 because of IM prefix!
+
+		// Use extension from filename if the name in the image node has a different extension:
+		std::string filepath = img->filepath;
+		size_t extensionIndex = filepath.find_last_of('.');
+		if (extensionIndex > 0) {
+			std::string extension = filepath.substr(extensionIndex);
+
+			if (name.find(extension) != name.length() - extension.length()) {
+				name += extension;
+			}
+		}
+
+		int exported = m_texture_manager->addImage(name, 0, 0, imagePackedFile->packedfile->data, imagePackedFile->packedfile->size, ms::TextureFormat::RawFile, textureType);
+		setTextureHandler(exported);
+	}
+}
+
+void msblenMaterialsExportHelper::exportImageFromImageNode(ms::TextureType& textureType,
+	std::function<void(int textureId)> setTextureHandler,
+	bNode* sourceNode) const
+{
+	if (setTextureHandler) {
```

Comment:
Can you switch this to `if (!setTextureHandler){return;}` to reduce the indentation

Reply:
Done